### ISSUE TITLE
Rtx Support!

### ISSUE THREAD

**Issue by MrScautHD** [xCollateral/VulkanMod#193]
Vulkan support RTX

soo pls add it!

**Comment by ezamelczyk** [comment]
Feel free to make a pull request with your changes.

**Comment by thr3343** [comment]
Unfortunately getting RTX/RDNA2 support off the ground will not be trivial due to the crazy stuff needed in a Ray-Tracing enabled pipeline, which is drastically different than the pipeline setup used currently for rendering, and will require the mod to behave more like blender than Minecraft AFAIK.

A ray-Tracing enabled pipeline needs a TLAS and BLAS (Top and Bottom-Level Acceleration Structure), and you also need special shaders as well (Ray-Gen, Hit, Miss and so on), and IIRC you need both to use RTX acceleration. These also have Rebuild times just like BVHs in Blender.

**Comment by thr3343** [comment]
The other problem is that AFAIK there isn't anyone around that actually knows how to write these special Raytracing shaders properly, which is also likely why RTX support was never added.

I have had ideas about making an attempt to setup a RTX/Raytracing PR before, but as mentioned above I'm not a ShaderPack author and don't know how to optimise the RayGen shaders properly, so it likely wouldn't run very well or be particularly playable.

**Comment by ghost** [comment]
if you get it running, it will probably get a lot of attention, then youtubers will make videos, rtx coder will see them, and then help you, you could try if think you can get it to work

**Comment by ghost** [comment]
> Feel free to make a pull request with your changes.

as if not, he verry probably has no idea how to code rtx into minecraft, and he also verry probably knows that you can make a pull request
so this commend is extremely useless

**Comment by ezamelczyk** [comment]
> > Feel free to make a pull request with your changes.
> 
> as if not, he verry probably has no idea how to code rtx into minecraft, and he also verry probably knows that you can make a pull request so this commend is extremely useless

True. Poor response on my part but I just don't like the commanding tone of the initial request.

**Comment by thr3343** [comment]
That's OK, don't think anyone was bothered or offended in anyway. 

IMHO alot of stuff and performance issues haven't been fixed or added yet for a longish while due to many good reasons _(lack of time and/or devs, Bugs with GPU Drivers, Vanilla MC code being difficult to work with e.g.)_ which isn't xCollateral's or anyone's fault.

I don't mind trying to experiment with Ray/Path-tracing to do something different, _(The PRs I've done either have massive issues or ugly code which I've wasted far too much time on)_, however I am mostly an amateur who doesn't fully understand what they're doing, so I can't guarantee anything.

I also don't have the best hardware _(GTX 1080 Ti which supports VK_KHR_ray_tracing_pipeline but is an outdated Architecture, lacks Ray Queries and is slow with Path-Tracing)_, so If anything comes of it, it will likely be as a experiment/tech demo.

**Comment by thr3343** [comment]
Also you don't necessarily need to have RTX, as it will likely also/only work with [these GPUs](https://vulkan.gpuinfo.org/listdevicescoverage.php?extension=VK_KHR_ray_tracing_pipeline) and as long as you have Vulkan 1.2, anything else causes the driver to crash AFAIK.

**Comment by nonetrix** [comment]
Yep! RADV also supports `emulate_rt=1` on older cards which I assume just makes it use the CPU or some slower GPU based fall back instead, this also let's you play Minecraft RTX on Linux on older AMD cards, but the Education version instead. Normal bedrock isn't packaged as .exe unlike Education but off topic rant but sucks Wine doesn't support UWP sometimes lol :/

**Comment by goofyseeker311** [comment]
bump this. implementation should be vulkan rt, not rtx or dxr, also because lwjgl has direct bindings to vulkan/rt. rtx is effectively the native nvidia implementation version of the vulkan rt api. which is not dependent on gpu vendor. public api.

so if you choose and gpu can run vulkan rt, then it should, otherwise fall back to shader rt or shader raster. Yes I mean you should run plain vulkan rt, not mix with rt/raster. to get the simplest implementation of the vulkan hw ray tracing api.

any relatively modern gpu can run vulkan rt, at least in shader emulation mode, if nothing else. any very recent gpu can all run vulkan rt with hw ray tracing cores, including nvidia, amd and intel gpus.

also lwjgl github repo has comprehensive working examples of using the vulkan hw rt api. basically everything in rendering is offloaded to the vulkan rt api. including the acceleration structures the api generates. internally. on the gpu. also the default shaders are there. usage of vulkan rt with the normal vulkan api (vulkanfw) libs should be simple enough.

so if you want to do shader rt using gpu hw api instead then lwjgl vulkan rt recommended.

**Comment by goofyseeker311** [comment]
at least bsl style rt shaders should be made to work. then continue to the hw rt shader api support.
